src/matrixrtc/BasicEncryptionManager.ts

```diff
@@ -0,0 +1,237 @@
+    }
+
+    private async rolloutOutboundKey(): Promise<void> {
+        const hasExistingKey = this.outboundSession != null;
+
+        // Create a new key
+        const newOutboundKey: OutboundEncryptionSession = {
+            key: this.generateRandomKey(),
+            creationTS: Date.now(),
+            sharedWith: [],
+            keyId: this.nextKeyId(),
+        };
+
+        this.logger.info(`creating new outbound key index:${newOutboundKey.keyId}`);
+        // Set this new key has the current one
+        this.outboundSession = newOutboundKey;
+        this.needToRotateAgain = false;
+        const toShareWith = this.getMemberships();
+
+        try {
+            this.logger.trace(`Sending key...`);
+            await this.transport.sendKey(encodeBase64(newOutboundKey.key), newOutboundKey.keyId, toShareWith);
+            this.statistics.counters.roomEventEncryptionKeysSent += 1;
+            newOutboundKey.sharedWith = toShareWith.map((ms) => {
+                return {
+                    userId: ms.sender ?? "",
+                    deviceId: ms.deviceId ?? "",
+                };
+            });
+            this.logger.trace(
+                `key index:${newOutboundKey.keyId} sent to ${newOutboundKey.sharedWith.map((m) => `${m.userId}:${m.deviceId}`).join(",")}`,
+            );
+            if (!hasExistingKey) {
+                this.logger.trace(`Rollout immediately`);
+                // rollout immediately
+                this.onEncryptionKeysChanged(
+                    newOutboundKey.key,
+                    newOutboundKey.keyId,
+                    getParticipantId(this.userId, this.deviceId),
+                );
+            } else {
+                // Delay a bit using this key
+                const rolledOut = defer<void>();
+                this.logger.trace(`Delay Rollout...`);
+                setTimeout(() => {
+                    this.logger.trace(`...Delayed rollout of index:${newOutboundKey.keyId} `);
+                    // Start encrypting with that key now that there was time to distibute it
+                    this.onEncryptionKeysChanged(
+                        newOutboundKey.key,
+                        newOutboundKey.keyId,
+                        getParticipantId(this.userId, this.deviceId),
+                    );
+                    rolledOut.resolve();
+                }, 1000);
+                return rolledOut.promise;
+            }
+        } catch (err) {
+            this.logger.error(`Failed to rollout key`, err);
+        }
+    }
+
+    private nextKeyId(): number {
+        if (this.outboundSession) {
+            return (this.outboundSession!.keyId + 1) % 256;
+        }
+        return 0;
+    }
+
+    private generateRandomKey(): Uint8Array {
+        const key = new Uint8Array(16);
+        globalThis.crypto.getRandomValues(key);
+        return key;
+    }
+}
+
+const getParticipantId = (userId: string, deviceId: string): ParticipantId => `${userId}:${deviceId}`;
```

Comment:
maybe move this to `utils` now that we have such a file.
there are also logs where we do `${userId}:${deviceId}` which can use the utils method.